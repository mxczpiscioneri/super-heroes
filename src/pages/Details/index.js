import React, { Component } from 'react'
import PropTypes from 'prop-types'
import {
  ScrollView,
  ActivityIndicator,
} from 'react-native'
import {
  Background,
  Character,
  Links,
  MoreInfo,
} from '../../components'
import { generateUrlImage } from '../../utils'
import { getCharacters } from '../../services/api'
import {
  Info,
  Description,
} from './styles'


class Details extends Component {
  static navigationOptions = ({ navigation }) => {
    const { state } = navigation
    return {
      title: state.params.title || '',
    }
  }

  constructor(props) {
    super(props)

    this.state = {
      loading: true,
      character: {},
      comics: [],
      series: [],
      stories: [],
      events: [],
    }
  }

  async componentDidMount() {
    const { getParam, setParams } = this.props.navigation
    const characterId = getParam('characterId')

    const response = await getCharacters(`/${characterId}`)
    setParams({ title: response.results[0].name })
    this.setState({
      character: response.results[0],
      loading: false,
    })

    this.showMoreInfo(characterId)
  }

  /**
   * View more character information
   * @param {String} characterId - Character id
   */
  showMoreInfo = async characterId => {
    getCharacters(`/${characterId}/comics`)
      .then(data => {
        this.setState({ comics: data.results })
      })

    getCharacters(`/${characterId}/series`)
      .then(data => {
        this.setState({ series: data.results })
      })

    getCharacters(`/${characterId}/stories`)
      .then(data => {
        this.setState({ stories: data.results })
      })

    getCharacters(`/${characterId}/events`)
      .then(data => {
        this.setState({ events: data.results })
      })
  }

  render() {
    const {
      loading,
      character,
      comics,
      series,
      stories,
      events,
    } = this.state

    return (
      <Background>
        {!loading ?
          <ScrollView>
            <Character
              character={character}
              image={generateUrlImage(character.thumbnail.path, character.thumbnail.extension, 'detail')}
            />

            <Info>
              <Description>{character.description}</Description>
              <Links character={character} />
            </Info>

            <MoreInfo title='Comics' data={comics} />
            <MoreInfo title='Series' data={series} />
            <MoreInfo title='Stories' data={stories} />
            <MoreInfo title='Events' data={events} />
          </ScrollView> :
          <ActivityIndicator />
        }
      </Background>
    )
  }
}

Details.propTypes = {
  navigation: PropTypes.object.isRequired,
}

export default Details
