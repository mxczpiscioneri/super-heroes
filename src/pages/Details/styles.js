import Styled from 'styled-components/native'

export const Info = Styled.View`
  flex: 1;
  flexDirection: row
  margin: 16px;
`

export const Description = Styled.Text`
  fontSize: 20px;
  fontWeight: 500;
  textAlign: center;
`
