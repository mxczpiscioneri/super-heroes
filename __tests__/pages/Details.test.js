
import React from 'react'
import renderer from 'react-test-renderer'
import Details from '../../src/pages/Details'

const navigation = { navigate: jest.fn() }

it('renders Details', () => {
  const rendered = renderer.create(
    <Details navigation={navigation} />
  ).toJSON()
  expect(rendered).toMatchSnapshot()
})
