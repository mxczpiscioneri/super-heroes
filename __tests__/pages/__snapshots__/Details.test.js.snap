// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Details 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Array [
      Object {
        "height": "100%",
        "width": "100%",
      },
    ]
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../src/statics/images/background.jpg",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;
